tapeout/src/main/scala/transforms/macros/MacroCompiler.scala

```diff
@@ -0,0 +1,401 @@
+              /* Palmer: If we're expected to provide mask ports without a
+               * memory that actually has them then we can use the
+               * write enable port instead of the mask port. */
+              stmts += connectPorts(andAddrMatch(and(memWriteEnable, memMask)),
+                                    we, libPort.writeEnablePolarity)
+              stmts += connectPorts(andAddrMatch(memChipEnable), en, libPort.chipEnablePolarity)
+            case (None, Some(we), Some(en)) =>
+              // TODO
+              System.err println "cannot emulate multi-bit mask ports with write enable"
+              return None
+            case (None, None, None) =>
+              /* Palmer: There's nothing to do here since there aren't any
+               * ports to match up. */ 
+          }
+        }
+        // Cat macro outputs for selection
+        memPort.outputName match {
+          case Some(mem) if cats.nonEmpty =>
+            val name = s"${mem}_${i}"
+            stmts += DefNode(NoInfo, name, cat(cats.toSeq.reverse))
+            (outputs getOrElseUpdate (mem, ArrayBuffer[(Expression, Expression)]())) +=
+              (addrMatch -> WRef(name))
+          case _ =>
+        }
+      }
+    }
+    // Connect mem outputs
+    mem.ports foreach { port =>
+      port.outputName match {
+        case Some(mem) => outputs get mem match {
+          case Some(select) =>
+            val output = (select foldRight (zero: Expression)) {
+              case ((cond, tval), fval) => Mux(cond, tval, fval, fval.tpe) }
+            stmts += Connect(NoInfo, WRef(mem), output)
+          case None =>
+        }
+        case None =>
+      }
+    }
+
+    Some((mem.module(Block(stmts.toSeq)), lib.blackbox))
+  }
+
+  def run(c: Circuit): Circuit = {
+    val modules = (mems, libs) match {
+      case (Some(mems), Some(libs)) => (mems foldLeft c.modules){ (modules, mem) =>
+        val (best, cost) = (libs foldLeft (None: Option[(Module, ExtModule)], BigInt(Long.MaxValue))){
+          case ((best, area), lib) if mem.ports.size != lib.ports.size =>
+            /* Palmer: FIXME: This just assumes the Chisel and vendor ports are in the same
+             * order, but I'm starting with what actually gets generated. */
+            System.err println s"INFO: unable to compile ${mem.name} using ${lib.name} port count must match"
+            (best, area)
+          case ((best, area), lib) =>
+            /* Palmer: A quick cost function (that must be kept in sync with
+             * memory_cost()) that attempts to avoid compiling unncessary
+             * memories.  This is a lower bound on the cost of compiling a
+             * memory: it assumes 100% bit-cell utilization when mapping. */
+            // val cost = 100 * (mem.depth * mem.width) / (lib.depth * lib.width) +
+            //                  (mem.depth * mem.width)
+            // Donggyu: I re-define cost
```

Comment:
I didn't understand Palmer's cost model and couldn't translate it. I think my cost model is what it should do, considering the utilization and the number of macros.